clippy_utils/src/ty.rs

```diff
@@ -322,3 +322,22 @@ pub fn walk_ptrs_ty_depth(ty: Ty<'_>) -> (Ty<'_>, usize) {
     }
     inner(ty, 0)
 }
+
+/// Returns `true` if types `a` and `b` are same types having same `Const` generic args,
+/// otherwise returns `false`
+pub fn same_type_and_consts(a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
+    match (&a.kind(), &b.kind()) {
+        (&ty::Adt(did_a, substs_a), &ty::Adt(did_b, substs_b)) => {
+            if did_a != did_b {
+                return false;
+            }
+
+            substs_a
+                .types()
+                .zip(substs_b.types())
+                .all(|(a, b)| same_type_and_consts(a, b))
+                && substs_a.consts().zip(substs_b.consts()).all(|(a, b)| a == b)
```

Comment:
It should be faster to iterate the generic args and use `unpack`. The `types` and `consts` methods will iterate and unpack each list an extra time.